org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/classfmt/ClassFileReader.java

```diff
@@ -162,8 +166,7 @@ public static ClassFileReader read(
 	java.util.zip.ZipEntry ze = zip.getEntry(filename);
 	if (ze == null)
 		return null;
-	byte classFileBytes[] = Util.getZipEntryByteContent(ze, zip);
-	ClassFileReader classFileReader = new ClassFileReader(classFileBytes, filename.toCharArray());
+	ClassFileReader classFileReader = Util.getZipEntryClassFile(zip.getName(), filename);
```

Comment:
Please help me understand, why we can't use the `ze`, after it has already been retrieved successfully?
Passing the URI into the `ClassFileReader` should be possible without `getZipEntryClassFile()`, no? Or is the encoding of an URI relative to a zip file the problem?

If opening a `ZipFileSystem` is a potential performance issue, I'd think an URI like `jar:file:///tmp/eclipse-temp/lib/apttestprocessors8.jar!/org/eclipse/jdt/compiler/apt/tests/annotations/Type.class` can be inexpensively constructed by simple string operations?

In tests, this seemed to work:

```
	URI uri = URI.create("jar:"+Paths.get(zip.getName()).toUri()+"!/"+filename);
```
Does it lack any fine points of URI encoding?